grpc-gcp/src/main/java/com/google/cloud/grpc/multiendpoint/MultiEndpoint.java

```diff
@@ -0,0 +1,201 @@
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+
+/**
+ * MultiEndpoint holds a list of endpoints, tracks their availability and defines the current
+ * endpoint. An endpoint has a priority defined by its position in the list (first item has top
+ * priority). MultiEndpoint returns top priority endpoint that is available as current. If no
+ * endpoint is available, MultiEndpoint returns the top priority endpoint.
+ *
+ * <p>Sometimes switching between endpoints can be costly, and it is worth to wait for some time
+ * after current endpoint becomes unavailable. For this case, use {@link
+ * Builder#withRecoveryTimeout} to set the recovery timeout. MultiEndpoint will keep the current
+ * endpoint for up to recovery timeout after it became unavailable to give it some time to recover.
+ *
+ * <p>The list of endpoints can be changed at any time with {@link #setEndpoints} method.
+ * MultiEndpoint will preserve endpoints' state and update their priority according to their new
+ * positions.
+ *
+ * <p>The initial state of endpoint is "unavailable" or "recovering" if using recovery timeout.
+ */
+@CheckReturnValue
+public final class MultiEndpoint {
+  @GuardedBy("this")
+  private final Map<String, Endpoint> endpointsMap = new HashMap<>();
+
+  @GuardedBy("this")
+  private String currentId;
+
+  private final Duration recoveryTimeout;
+  private final boolean recoveryEnabled;
+
+  private final ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1);
+
+  private MultiEndpoint(Builder builder) {
+    this.recoveryTimeout = builder.recoveryTimeout;
+    this.recoveryEnabled =
+        !builder.recoveryTimeout.isNegative() && !builder.recoveryTimeout.isZero();
+    this.setEndpoints(builder.endpoints);
+  }
+
+  /** Builder for MultiEndpoint. */
+  public static final class Builder {
+    private final List<String> endpoints;
+    private Duration recoveryTimeout = Duration.ZERO;
+
+    public Builder(List<String> endpoints) {
+      Preconditions.checkNotNull(endpoints);
+      Preconditions.checkArgument(!endpoints.isEmpty(), "Endpoints list must not be empty.");
+      this.endpoints = endpoints;
+    }
+
+    /**
+     * MultiEndpoint will keep the current endpoint for up to recovery timeout after it became
+     * unavailable to give it some time to recover.
+     */
+    public Builder withRecoveryTimeout(Duration timeout) {
+      Preconditions.checkNotNull(timeout);
+      this.recoveryTimeout = timeout;
+      return this;
+    }
+
+    public MultiEndpoint build() {
+      return new MultiEndpoint(this);
+    }
+  }
+
+  /**
+   * Returns current endpoint id.
+   *
+   * <p>Note that the read is not synchronized and in case of a race condition there is a chance of
```

Comment:
This is because GcpMultiEndpointChannel provides a healthy top-priority endpoint on a best-effort basis. We already have some delay when receiving status updates from channels, so it is expected that occasionally a few RPC (out of thousands) could go to an unavailable or less-prioritized endpoint. But for high QPS clients it is not a problem (solved using retries) and for low QPS clients the chance of picking a suboptimal endpoint is very low.

I think the cost of synchronizing on getting the current endpoint (which happens on every RPC call) is higher than getting a suboptimal endpoint in edge cases.